File: backend/internal/models/user_task_progress.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

// UserTaskProgress stores per-user progress for a task.
type UserTaskProgress struct {
	ID           bson.ObjectID `bson:"_id,omitempty" json:"id"`
	UserID       bson.ObjectID `bson:"userId" json:"userId"`
	TaskID       bson.ObjectID `bson:"taskId" json:"taskId"`
	Attempts     int           `bson:"attempts" json:"attempts"`
	CompletedAt  *time.Time    `bson:"completedAt,omitempty" json:"completedAt,omitempty"`
	PointsEarned int           `bson:"pointsEarned" json:"pointsEarned"`
	CreatedAt    time.Time     `bson:"createdAt" json:"createdAt"`
	UpdatedAt    time.Time     `bson:"updatedAt" json:"updatedAt"`
}

File: server/src/models/UserTaskProgress.js
import mongoose from 'mongoose'

const userTaskProgressSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'User ID is required'],
      index: true,
    },
    taskId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Task',
      required: [true, 'Task ID is required'],
      index: true,
    },
    attempts: {
      type: Number,
      default: 0,
    },
    completedAt: {
      type: Date,
      default: null,
    },
    pointsEarned: {
      type: Number,
      default: 0,
    },
  },
  {
    timestamps: true,
    collection: 'user_task_progress',
  },
)

userTaskProgressSchema.index({ userId: 1, taskId: 1 }, { unique: true })
userTaskProgressSchema.index({ completedAt: -1 })

const UserTaskProgress = mongoose.model(
  'UserTaskProgress',
  userTaskProgressSchema,
)

export default UserTaskProgress
